Engineering mode

skoona opened the pull request:
Added support for the commands and formatting of Engineering Data.  Did not add getters to read related data at this time.

skoona committed:
Formatting text

skoona committed:
Enable Commands Acks for Engineering Mode

skoona committed:
Adding Engineering Supports

skoona committed:
First Good Compile after adding engineering mode

skoona committed:
Update Engineering mode commands

skoona committed:
Formatting Engineering data

skoona committed:
Add the example script used to produce logfile, attached later

skoona committed:
Fix buffer size parsing

skoona committed:
Fix buffer size parsing

skoona committed:
Both engineering commands added in original style

skoona committed:
Realign constants and remove unused methods

skoona committed:
Minor Cleanups

skoona committed:
Minor Cleanups to line formats

skoona committed:
added Set serial baud Rate

skoona committed:
Updated the library APIs to support less chatty interface

skoona committed:
Updated the library APIs to support less chatty interface

skoona committed:
Fix minor syntax error

skoona reviewed (commented):
Assumes a user would use the GPIO with a few APIs for conformation

skoona committed:
Add some utility methods and ensure enter_config.. issued before every command per spec.

skoona committed:
Turn off debug flags

skoona committed:
Update Example program and enhance ,begin() to include configuraiton info vs firmware

skoona committed:
Update begin delays to allow for device restart.

skoona committed:
Implement detection_distance_, which was omitted

skoona committed:
Update Exaample to include detection distance

skoona committed:
Update the example and document debug flags

skoona committed:
Apply observed coding standards

skoona commented:
PR is complete at this point

skoona committed:
Fix Bug around line 83 where debug_uart was unprotected on no-respone.

ncmreynolds reviewed (dismissed):
I don't use PlatformIO myself so am approving this untested.

skoona committed:
Merge branch 'main' into engineering_mode

skoona commented:
Resolved merge conflict

ncmreynolds commented:
Hi, I've checked out a copy of this and had a look. I don't think I can merge it with printf in use as this will cause it to not compile for some older but still common Arduino targets, especially AVR.

One of my own use cases is to use an old AVR Arduino as an I2C peripheral with a LD2410 connected where I'm already out of UARTs on the main microcontroller in a project.

You've tidied up a lot of my messy repetition though, thanks.

skoona committed:
Removed printf from library and updated examples and implemented all command strategies.

skoona commented:
This last commit removed the printf and snprintf functions.  I implemented all the commands from the LD2410 specification doc, and added some comments.  The specs say command-mode once entered will support multiple follow on commands unless a command ends in error.  I added user methods to begin/end command mode to allow this multiple command feature.

The advanced example is essentially my Skn2410 program which uses UDP to send reading to SerialStudio and allow configuration of the LD2410 over UDP. To test configure path you could use this command: "nc -u <ip of esp> 8091", and issue "help" to get the menu of commands available.

The library code itself does not use printf or snprintf functions (there were only 4), so we should be fine. 

Let me know if you see any other issues.

ncmreynolds commented:
Thanks, I'll grab this and test it with a sensor.

On Wed, 7 Dec 2022 at 14:26, James Scott, Jr ***@***.***>
wrote:

> This last commit removed the printf and snprintf functions. I implemented
> all the commands from the LD2410 specification doc, and added some
> comments. The specs say command-mode once entered will support multiple
> follow on commands unless a command ends in error. I added user methods to
> begin/end command mode to allow this multiple command feature.
>
> The advanced example is essentially my Skn2410 program which uses UDP to
> send reading to SerialStudio and allow configuration of the LD2410 over
> UDP. To test configure path you could use this command: "nc -u 8091", and
> issue "help" to get the menu of commands available.
>
> The library code itself does not use printf or snprintf functions (there
> were only 4), so we should be fine.
>
> Let me know if you see any other issues.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ncmreynolds/ld2410/pull/3#issuecomment-1341045472>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOP5GD23BG4RWFQSSKJPL3WMCNB7ANCNFSM6AAAAAASQDBOCA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Nick Reynolds - 950SM|620SC

skoona committed:
Fix cmdFirmwareVersion(), concat were wrong and was return ptr to local var

skoona committed:
Add device info and removed menu for engineering mode

skoona committed:
Tweak the titles in advanced example and added an ESP reboot option.

skoona committed:
Update Engineering extra Value at end of reporting data, was overlapped with idle_time

Commit:
Qualified serial write() method to quiet compiler

skoona committed:
Update menu text

PhilF66 commented:
I've been following this with interest.  One of the things that has troubled me is the parsing of the "energy level" data when Engineering Mode is set.  The available documentation suggests that there are 8 "gates", and one would glean that the 8 values for motion are immediately followed by the 8 values for stationary targets when parsing the relevant part of the data frame.  The current code does exactly this, starting at position 19 within the frame.  It populates the array "movement_distance_gate_energy" and immediately carries on to populate "static_distance_gate_engergy" (yes, it's spelled that way) where it left off.  This just didn't look right, as I'm focused on the stationary "energy" values,  and it was particularly troubling that there are a couple of values in there that always appear to be zero.

Then a blinding flash of the obvious.  If you look at the way the data is presented by the LD2410 tool provided by Hi-Link, the gates are numbered 0-8.  NINE values.  There is activity shown in all 9 gates of the "motion" graph while running, and yet only gates 2 through 8 show anything on the "stationary" side.  Gates 0 and 1 always show 0, and the default factory sensitivity levels for these 2 gates is always set to zero, as well, which is interesting.  Anyways, if you either treat these arrays as having 9 values, or simply skip a byte when moving on to parse the stationary data, you get results that are consistent with what the Hi-Link utility shows.
![image](https://user-images.githubusercontent.com/110070300/206951123-fbd6b1b6-56b2-47f8-88d2-641f3422cc1a.png)

PhilF66 commented:
I had left an earlier remark about the parsing of the "energy" data which I realize was the result of my taking a trip down the wrong rabbit hole.  Disregard, should you have already seen it.  This is great work!

ncmreynolds commented:
Have you finished making changes to the pull request for now? I've not
properly reviewed it yet as you seem to keep making small changes.

On Mon, 12 Dec 2022 at 03:42, Phil French ***@***.***> wrote:

> I had left an earlier remark about the parsing of the "energy" data which
> I realize was the result of my taking a trip down the wrong rabbit hole.
> Disregard, should you have already seen it. This is great work!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ncmreynolds/ld2410/pull/3#issuecomment-1345826191>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOP5GEXZ47MA6HV376GBADWM2NJ3ANCNFSM6AAAAAASQDBOCA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Nick Reynolds - 950SM|620SC

skoona commented:
Yes, I'm finished

skoona committed:
Target type was over qualified

eried commented:
hi! basically this will expose the energy levels of the 9 gates that are 75 cms away yes? @skoona 

@ncmreynolds  any reason this was not merged? found something wrong? (I wonder if I should use this version of try to work on the original)

ncmreynolds commented:
> hi! basically this will expose the energy levels of the 9 gates that are 75 cms away yes? @skoona
> 
> @ncmreynolds any reason this was not merged? found something wrong? (I wonder if I should use this version of try to work on the original)

This PR has a massive amount of changes in one go and I've struggled to find time to meaningfully review it because of this.

I've just checked out this PR and tested it with a sensor and it seems to report incorrect values. In the basic example it constantly shows a stationary target at 0cm. Only occasionally shows a believable reading for the stationary target. Moving targets seem OK.

I can't merge this as it stands.

ncmreynolds commented:
There is lots of good work in this PR and I am likely to manually add some of it to the main branch, but as mentioned I can't merge it as it stands.

eried commented:
yeah I understand. I feel the documentation of the protocol is quite cryptic for this mode, it would be so easy for them to just explain it better with an example (I know there is an example but not labelled)

joba-1 reviewed (changes requested):
I just started using this sensor and found the library very helpful. Btw: I use the raspberry pico w. No changes required, it just worked.
But now that the feet are wet, some deficiencies showed up: I wrongly assumed methods *TargetDetected() considered gate energy threshold levels, but they don't.
This probably means I will need engineering mode. So here I am:
I fully understand this PR is hard to review - I just tried. It is too big, diffs get out of sync.
Other than that, I agree: lots of useful stuff in here, I like the style.
One change would be required: the basic example will not work because some attributes are now protected (they are available as getters now).
Hope you don't mind...

joba-1 commented on src/ld2410.cpp:
```diff
@@ -444,355 +393,366 @@ bool ld2410::parse_data_frame_()
+
+bool ld2410::parse_command_frame_()
+{
+	uint16_t intra_frame_data_length_ = serial_to_int_(4); //radar_data_frame_[4] + (radar_data_frame_[5] << 8);
```
move definition of intra_frame_data_length_ inside following #ifdef (unused elsewhere)

joba-1 commented on src/ld2410.cpp:
```diff
@@ -23,6 +23,54 @@ ld2410::~ld2410()	//Destructor function
 
+uint16_t ld2410::serial_to_int_(uint8_t index)
+{
+    return (int16_t) radar_data_frame_[index++] + (radar_data_frame_[index] << 8);
```
my compiler complains about using index++ and index in the same expression -> undefined (...which is evaluated first)

joba-1 commented on src/ld2410.cpp:
```diff
@@ -31,21 +79,49 @@ bool ld2410::begin(Stream &radarStream, bool waitForRadar)	{
 	}
 	if(waitForRadar)
 	{
+		if(requestRestart()) {
```
My sensor does not survive this restart. It gets response OK, but next command requestCurrentConfiguration() always fails. If I just increase the delay to 2500, then every second host restart works.
